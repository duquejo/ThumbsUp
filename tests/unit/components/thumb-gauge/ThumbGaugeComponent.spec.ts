import { describe, it, expect, beforeEach, vi } from 'vitest';
import { mount, shallowMount } from '@vue/test-utils';
import ThumbGaugeComponentVue from '@/components/thumb-gauge/ThumbGaugeComponent.vue';

describe('+ ThumbGaugeComponent unit tests', () => {

  const properties = {
    votes: {
      positive: 70,
      negative: 32
    }
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('should render properly', () => {
    const wrapper = shallowMount(ThumbGaugeComponentVue);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render with the default props', () => {
    const wrapper = shallowMount(ThumbGaugeComponentVue);
    
    const thumbsValues = wrapper.findAll('.icon-values span');
    const thumbsBackground = wrapper.findAll('.icon-percentage');

    expect(thumbsValues[0].text()).toBe('0%');
    expect(thumbsValues[1].text()).toBe('0%');

    expect(thumbsBackground[0].attributes('style')).toBe('width: 0%;');
    expect(thumbsBackground[1].attributes('style')).toBe('width: 0%;');
  });

  it('should render with the given props', () => {
    const wrapper = shallowMount(ThumbGaugeComponentVue, {
      props: { ...properties }
    });
    expect(wrapper.html()).toMatchSnapshot();

    const thumbsValues = wrapper.findAll('.icon-values span');
    const thumbsBackground = wrapper.findAll('.icon-percentage');

    expect(thumbsValues[0].text()).not.toBe('0%');
    expect(thumbsValues[1].text()).not.toBe('0%');

    expect(thumbsBackground[0].attributes('style')).not.toBe('width: 0%;');
    expect(thumbsBackground[1].attributes('style')).not.toBe('width: 0%;');
  });

  it('should calculate correct percentages for positive and negative votes', async () => {
    const wrapper = mount(ThumbGaugeComponentVue, {
      props: { ...properties },
    });

    const thumbsValues = wrapper.findAll('.icon-values span');
    const thumbsBackground = wrapper.findAll('.icon-percentage');

    const totalOfVotes = properties.votes.positive + properties.votes.negative;

    const positiveCount = Math.round((properties.votes.positive / totalOfVotes) * 100);
    const negativeCount = Math.round((properties.votes.negative / totalOfVotes) * 100);

    expect(thumbsValues[0].text()).toBe(`${positiveCount}%`);
    expect(thumbsValues[1].text()).toBe(`${negativeCount}%`);

    expect(thumbsBackground[0].attributes('style')).toBe(`width: ${positiveCount}%;`);
    expect(thumbsBackground[1].attributes('style')).toBe(`width: ${negativeCount}%;`);
  });
});
