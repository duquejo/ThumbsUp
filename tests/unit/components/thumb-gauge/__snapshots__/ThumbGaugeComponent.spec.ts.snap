// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`+ ThumbGaugeComponent unit tests > should render properly 1`] = `
"<div data-v-a282580d="" class="card__thumbs-percentage">
  <div data-v-a282580d="" class="icon-values"><span data-v-a282580d=""><img data-v-a282580d="" src="/assets/img/thumbs-up.svg" alt="thumbs down">0%</span><span data-v-a282580d="">0%<img data-v-a282580d="" src="/assets/img/thumbs-down.svg" alt="thumbs down"></span></div>
  <div data-v-a282580d="" class="icon-background"><span data-v-a282580d="" class="icon-percentage" aria-label="thumbs up" style="width: 0%;"></span><span data-v-a282580d="" class="icon-percentage" aria-label="thumbs down" style="width: 0%;"></span></div>
</div>"
`;

exports[`+ ThumbGaugeComponent unit tests > should render with the given props 1`] = `
"<div data-v-a282580d="" class="card__thumbs-percentage">
  <div data-v-a282580d="" class="icon-values"><span data-v-a282580d=""><img data-v-a282580d="" src="/assets/img/thumbs-up.svg" alt="thumbs down">69%</span><span data-v-a282580d="">31%<img data-v-a282580d="" src="/assets/img/thumbs-down.svg" alt="thumbs down"></span></div>
  <div data-v-a282580d="" class="icon-background"><span data-v-a282580d="" class="icon-percentage" aria-label="thumbs up" style="width: 69%;"></span><span data-v-a282580d="" class="icon-percentage" aria-label="thumbs down" style="width: 31%;"></span></div>
</div>"
`;
